### ISSUE TITLE
[REQUEST] Automatic generation of types from tests

### ISSUE THREAD

**Issue by moritonal** [rswag/rswag#643]
## Is your feature request related to a problem? Please describe.
Having inherited a massive project that has no spec tests and along with tending to disagree that a testing framework is always the best place for the API contract to exist, I would prefer if I could automatically generate my spec from the inputs & outputs of my controllers rather than strict tests.

## Describe the solution you'd like
The ability to write brief tests that test controllers, and then generate an OpenAPI document from the interactions.

## What support could we give you, so you could implement this yourself?
I have mostly successfully managed to implement this as part of a `include_context` option. Firstly, may I say thanks for your concise style of coding that made this easy.

The result looks like this:

```
get "Get Company by ID" do

    tags 'Companies'
    consumes 'application/json'
    produces 'application/json'

    parameter name: :id, in: :path, type: :integer

    response '200', 'Company found' do

        let(:id) { @company.id }

        include_context 'with auto-generated response'

        run_test!
    end
end

put "Update Company by ID" do
  tags 'Companies'
  consumes 'application/json'
  produces 'application/json'

  parameter name: :id, in: :path, type: :integer
  parameter name: :company, in: :body, type: :object, required: true, schema: {
    type: :object,
    properties: {
      name: { type: :string }
    }
  }

  response '200', 'Company updated' do

    let(:id) { @company.id }
    let(:company) { @company }

    include_context "with auto-generated request params"
    include_context 'with auto-generated response'
    run_test!
  end
end
```

## Describe alternatives you've considered
I attempted controller annotation but that lead to lots of bloat, and writing this I realise there must be a way to use `annotate` to get the types from the schema and then use that in some way. This would however suffer given the middleware's affects on the API.

## Additional context
My question, is what is the best course of action here. I'm new to the community so don't know the process. Firstly does this even align with RSwag's design? Does this feature sound like a library I should put together and release separately, offer to PR into one of the projects here, or just blog about?

## Relates to which version of OAS (OpenAPI Specification)
- [ ] OAS2
- [x] OAS3
- [x] OAS3.1

**Comment by pboling** [comment]
Wow Stale Bot is harsh, and not cool.

**Comment by moritonal** [comment]
It's okay. I didn't drive this forward given I didn't have much time to champion it and now have some more context having used it in earnest for a year.

Have to kinda eat my hat and say I got lost in the code for this one. Whilst it's helpful as an initial way to roughly map out your initial API I migrated away from this approach completely.  The idea was that it'd be an easy test-driven way to map types, but in reality it ended up being a slow & fragile way to update a couple fields that you already knew the answer to. Moreover, it had to make some very aggressive assumptions about the names of things and how various inheritances worked, leading to a poorer quality Swagger output.

So helpful as a starting point for producing an initial Swagger file, but not fit for wider use. Feel free to close and I'll try write up the code somewhere useful.
